### ISSUE TITLE
Support source files for external functions

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#2145]
**Reported by beutlich on 4 Jan 2017 14:12 UTC**
This is a split ticket of Bernhard's useful https://github.com/modelica/ModelicaSpecification/issues/1088#issuecomment-435637977 to better support external function compilation/build, particularly
 * (multiple) source files
 * compiler flags (CFLAGS)
 * preprocessor defines (CPPFLAGS)

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/2145

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 12 Oct 2017 10:02 UTC**
To discuss at meeting. Some should be possible to trivially close.

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 19 Oct 2017 07:53 UTC**
Seems that a generic autotools (or similarly) would be a better solution - since we need compiler-specific CFLAGS etc.

Francesco: Standardize one level up - a box with API to open it.
Related to package management.
Would solve problem with C-functions in MSL.
(The automake-idea is possible duplicate of another ticket.)

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 25 Oct 2017 07:40 UTC**
Based on the conclusions it seems this should be merged with #1668. 
I thus close this as a duplicate.

**Comment by modelica-trac-importer** [comment]
**Modified by beutlich on 25 Oct 2017 07:45 UTC**

**Comment by beutlich** [comment]
Together with @hubertus65:

Because we need to suport source code FMUs better, we need to have sources to all external dependencies specified inside MSL and other Modelica libraries in a standardized way, so that FMU exporting Modelica tools can include them in the created FMU.

The minimal support we see is the additional support of source files and source directories as annotations. There is no strong need to support a full build system (with flags, preprocessor defines as propsed by #1668) now, since FMI also will not define an entire build system.

**Comment by beutlich** [comment]
@HansOlsson Can you put this also on the topics for Modelica 3.5?

**Comment by beutlich** [comment]
An alternative proposal to just files/directories is to reference a zip archive containing all the sources. This is currently supported by SimulationX.

**Comment by sjoelund** [comment]
That FMI decides to not support a build system even though it should be a 100% requirement is not really a good reason to skip it. You will need to set preprocessor flags and enable/disable parts in general. You can't just put the HDF5 sources in the FMU... they're too big. You want to look for HDF5 and if it exists, link against it.

**Comment by andreas-junghanns** [comment]
To clarify "support a build system 100%": FMI will support the build process with declarations of source language and version, compile flags and additional information (e.g. dependencies). There is no way to support all unknown build systems for any target platform and compiler/linker tool chain. We have to rely on the language semantics being sufficient to allow *most* build systems to make a reasonable guess how to create a binary from the provided sources (statically and dynamically linkable).

**Comment by HansOlsson** [comment]
> An alternative proposal to just files/directories is to reference a zip archive containing all the sources. This is currently supported by SimulationX.

The proposal in the other ticket was:
One proposal to solve this is to have an annotation SourceDirectory (similary as LibraryDirectory and IncludeDirectory) possibly with default "modelica://LibraryName/Resources/Source". That can then be used when building a source code FMU by first copying the contents of that that source-directory to the FMU.

Having a zip-directory instead of a normal directory should work I guess. And as far as I know build-processes normally rely on some variant of makefiles and if they are in the directory it might just work.
We can also specify the kind of makefiles (or similarly) that are supported for build-systems - but at least it seems we are on a similar page.

**Comment by sjoelund** [comment]
>  There is no way to support all unknown build systems for any target platform and compiler/linker tool chain.

Specifying that autotools or cmake is used and how to access it would simplify things though (pkg-config could work together with a custom build system for dependencies, but often does not exist for all libraries). I don't understand why we would want to specify our own system when we could just use something else that is known to work in (most) cases. I mean, cmake has been around for 19 years and still is not great (very hard to check which flags are needed to make libraries link; especially which flags are needed for static linking); it is the bare minimum that works though.

Note that if FMI does not specify how to use the makefiles, having them in the FMU does not quite help (like today where we have source-code FMUs and users don't read the instructions for how to compile them).

**Comment by svorkoetter** [comment]
I wouldn't want it to be restricted to a zip file, but I see no reason why a SourceDirectory annotation couldn't be used for both, being allowed to specify either a directory or a zip file.

On Thu, 07 Mar 2019 23:58:14 -0800
Hans Olsson <notifications@github.com> wrote:

> > An alternative proposal to just files/directories is to reference a
> > zip archive containing all the sources. This is currently supported
> > by SimulationX.
> 
> The proposal in the other ticket was:
> One proposal to solve this is to have an annotation SourceDirectory
> (similary as LibraryDirectory and IncludeDirectory) possibly with
> default "modelica://LibraryName/Resources/Source". That can then be
> used when building a source code FMU by first copying the contents of
> that that source-directory to the FMU.
> 
> Having a zip-directory instead of a normal directory should work I
> guess. And as far as I know build-processes normally rely on some
> variant of makefiles and if they are in the directory it might just
> work. We can also specify the kind of makefiles (or similarly) that
> are supported for build-systems - but at least it seems we are on a
> similar page.
> 
> -- 
> You are receiving this because you were assigned.
> Reply to this email directly or view it on GitHub:
> https://github.com/modelica/ModelicaSpecification/issues/2145#issuecomment-470839664


-- 
Stefan Vorkoetter <svorkoetter@maplesoft.com>

**Comment by HansOlsson** [comment]
Discussion at design meeting: seems good for next release.
Adding everything in the folder seems like a good start (specifying all individual files seems like too much work).

**Comment by beutlich** [comment]
From the Dymola 2020 Release Notes:

> Improved export of source code FMUs using external c-functions
>
> When exporting source code FMUs using external c-functions the source code of those functions will now be included.
> Specifically:
> * The contents of includeDirectory will be included under sources/include
> * The contents of `__Dymola_SourceDirectory` (default “…/Resources/Source”) will be included under sources/source

While not being complete, this can be seen as a starting point.

**Comment by HansOlsson** [comment]
Proposal:

For 3.5 we could add:
The sourceDirectory (default “…/Resources/Source”) gives the location for source files. It is not specified how they are built.

Or we can completely postpone it.

**Comment by beutlich** [comment]
* Does this mean to introduce the SourceDirectory annotation?
* Would this also imply to rename Resources/C-Sources to Resources/Source in the MSL (in case we want to utilize the default source directory)?

**Comment by HansOlsson** [comment]
> * Does this mean to introduce the SourceDirectory annotation?

Yes.

> * Would this also imply to rename Resources/C-Sources to Resources/Source in the MSL (in case we want to utilize the default source directory)?

Yes, or we could change the default to be Resouces/C-Sources  - either is fine by me.

**Comment by HansOlsson** [comment]
I realized that I missed this at the meeting, so let's do a poll (one week).

There are three options:
As above (rocket): 

The sourceDirectory (default “…/Resources/Source”) gives the location for source files.
It is not specified how they are built.

Modified to be MSL compatible (eyes):

The sourceDirectory (default “…/Resources/C-Sources”) gives the location for source files.
It is not specified how they are built.

No change (-1).
--
Pick symbols under the smiley close to the top-right corner in GitHub-message.
You can be in favor of zero, one, or both of the first two, or the last one.

**Comment by beutlich** [comment]
> Modified to be MSL compatible (eyes):

It is not consistent in MSL v4.0.0 anyway, since both sources and include headers are distributed as Resources/C-Sources. In case the include headers will be moved to Resources/Include we can also change the sources directory as we like.

**Comment by henrikt-ma** [comment]
(I didn't vote for no change because I don't like the idea.  I voted for no change because I think the idea is important enough to deserve a complete design before being introduced.  In particular, I'd like to make sure that there will be a system for building binary one or more binary libraries from the distributed source files, libraries that should linked by means of the `Library` annotation in external functions.)

**Comment by beutlich** [comment]
* What this change is about: In case of FMU export from Modelica all files/folders with the annotated sources directory will be included in the FMU.
* What it is not about: Building sources and add as linker dependency. We still have #1668 left open for this.

**Comment by beutlich** [comment]
By the way, FMI v2.0.2 introduced the extra directory (as another place to store metadata). Should we also add another annotation on external functions to support the extra directory?

**Comment by henrikt-ma** [comment]
> * What this change is about: In case of FMU export from Modelica all files/folders with the annotated sources directory will be included in the FMU.

OK, but the proposal still seems to lack some important details.  Is the idea that every Modelica external function wrapper should have a `SoruceDirectory` annotation pointing to a directory containing the files needed to provide the external implementation?  To me that doesn't really sound like the kind of direction that would fit well into a system for building sources that can be linked with normal simulations.  On the other hand, if this is really about a feature which is only meant to serve FMU export, maybe this should be reflected more clearly by the annotation's name?

**Comment by sjoelund** [comment]
@henrikt-ma I guess you have a point there. What is needed is actually a package level mapping from Library to the set of source files / etc (or pointer to another Modelica library that actually provides the library; many libraries use MSL C tables / etc in their external functions) needed to compile this library.

It would be too much to duplicate this for every function.

**Comment by HansOlsson** [comment]
> (I didn't vote for no change because I don't like the idea. I voted for no change because I think the idea is important enough to deserve a complete design before being introduced. In particular, I'd like to make sure that there will be a system for building one or more binary libraries from the distributed source files, libraries that should linked by means of the `Library` annotation in external functions.)

I can understand that sentiment.
It's just intended as a start, and not super-useful at this stage - but on the other hand I don't see how having a reference to the directory of source files will be a problem.

**Comment by henrikt-ma** [comment]
> It's just intended as a start, and not super-useful at this stage - but on the other hand I don't see how having a reference to the directory of source files will be a problem.

I'm concerned that the introduction of a half-finished `SourceDirectory` might still get widespread use that will get in the way the day when we want to fix the Modelica library build problem.

Without trying to solve the problem of how to specify detailed build instructions, just imagine that there would be a way for a Modelica library to specify where to find the sources for building a library of external functions:
```
package MyLib
  annotation(Build(Library(Name = "mylibfoo", SourceDirectory = "Resources/source/foo")));
end MyLib;
```
Then an external function would just need:
```
function MyExtFun
  …
external "C"
  … annotation(Library = "mylibfoo"`, `Include = "include \"foo.h\"");
end MyExtFun;
```

When exporting a source code FMU making use of `MyExtFun`, the necessary information would be available by looking up the `Library` annotation to see that the corresponding sources are found in _Resources/source/foo_.

While this is the kind of use of `SourceDirectory` I'd like to see, my impression of the current proposal is that it would be used quite differently.

**Comment by beutlich** [comment]
Waht about renaming to "ExportDirectory"?

**Comment by HansOlsson** [comment]
> Waht about renaming to "ExportDirectory"? This makes it clearer that it is not meant for building the external lib.

To me that would be a bit confusing as I would think it would wonder if it is a directory for exporting to instead of a directory used when exporting from.

**Comment by HansOlsson** [comment]
> > It's just intended as a start, and not super-useful at this stage - but on the other hand I don't see how having a reference to the directory of source files will be a problem.
> 
> I'm concerned that the introduction of a half-finished `SourceDirectory` might still get widespread use that will get in the way the day when we want to fix the Modelica library build problem.
> 
> Without trying to solve the problem of how to specify detailed build instructions, just imagine that there would be a way for a Modelica library to specify where to find the sources for building a library of external functions:
> 
> ```
> package MyLib
>   annotation(Build(Library(Name = "mylibfoo", SourceDirectory = "Resources/source/foo")));
> end MyLib;
> ```
> 
> Then an external function would just need:
> 
> ```
> function MyExtFun
>   …
> external "C"
>   … annotation(Library = "mylibfoo"`, `Include = "include \"foo.h\"");
> end MyExtFun;
> ```

I would view that as reducing duplicated information, which I view as a separate topic.

Having Library, Include etc for every function is a bit odd in libraries with lots of functions (especially if the library line contains several libraries) - and obviously that also applies to Build. But on the other hand that is convenient for users that can find everything listed at one function without the need for tool-support for tracking down the information.

So if we stick with the current duplication for the Build-information the code above would be:

```
function MyExtFun
  …
external "C"
  … annotation(Library = "mylibfoo", Include = "include \"foo.h\"", SourceDirectory="Resources/source/foo", Build(...));
end MyExtFun;
```
(where Build contains some specific Build-parts for building the library that are yet to be specified.)

And if we decide to allow gathering the information hierarchically one could then do:
```
package MyLib
  annotation(Build(Library(Include = "include \"foo.h\", Name = "mylibfoo", SourceDirectory = "Resources/source/foo", ...)));

  function MyExtFun
    …
  external "C";
  end MyExtFun;
end MyLib;
```
(possibly called something other than Build and with something special in ...). Thus I don't find a problem with adding this at the moment.

**Comment by henrikt-ma** [comment]
> I would view that as reducing duplicated information, which I view as a separate topic.

In my opinion, it's not a good habit to design new features that are known to cause duplication of information when used.  Changing design is always painful (like first saying that every external function should have `SourceDirectory`, and next version say that this information belongs at the library level instead), so one should try to get it right from start.

> Having Library, Include etc for every function is a bit odd in libraries with lots of functions (especially if the library line contains several libraries) - and obviously that also applies to Build. But on the other hand that is convenient for users that can find everything listed at one function without the need for tool-support for tracking down the information.

To me, having both `Library` and `Include` are a natural part of every external function, and not something I consider duplicated information; a large Modelica library may well contain external functions with implementations spread out in different binary libraries.  I also doubt that users would find it challenging to go and find out based on which sources a library is built.

Anyway, these are just my opinions.  If others don't find them compelling, there's no need to convince me; I'll just have to accept a different outcome of the poll.

**Comment by HansOlsson** [comment]
Poll results:
No change(-1): 2 votes: maltelenz and henrikt-ma 
Abstain(heart): 1 vote: qlambert-pro 
Original variant of addition (rocket): 6 votes: eshmoylova, beutlich, HansOlsson, hubertus65, dietmarw,  and adrpo 
Modified "MSL compatible" variant (eyes): 3 votes:  eshmoylova, hubertus65, and casella 
(Some were in favor of both variants of the addition.)

It seems clear that the original variant of the addition won.

**Comment by beutlich** [comment]
Just a question: Could we also update the example in 12.9.4 to make use of the SourceDirectory annotation. I was thinking of moving ExternalFunc3.c from Include to Sources, which would require to change the IncludeDirectory of ExternalFunc3, too.

Reopening to just not forget it.

**Comment by henrikt-ma** [comment]
IMHO, the less we make use of this half-finished feature, the better.

**Comment by beutlich** [comment]
If it is in the spec, it will be used for sure. I am also considering to make use of it in MSL.

**Comment by sjoelund** [comment]
The include annotation is in a pretty weird area. It's not necessary to add an include for the prototypes since the specification says exactly what they will look like. And tools that do not compile code (only performing dynamic loading of libraries), cannot use it.

It would make more sense to have an InlineSource feature in that case. Because it is a useful thing to have.

**Comment by beutlich** [comment]
I fully agree. I really would like to get rid of the inlcusion of C soure files, which I consider a bad design, too. Unfortunately, it is the most widely used feature of using external code, since the modeller does not need to consider compiling or linking at all. Having an InlineSource annotation sounds like a cleaner design to me.

**Comment by henrikt-ma** [comment]
Of course we should get rid of inclusion of C source files, but the ideal way to get there in my opinion would be to specify a library to link with instead.  Of course, a complete solution for this depends on the non-trivial #1668, but I'd even find an incomplete solution for this better than introducing a `SourceDirectory` on every external function that will just become redundancy burden once #1668 has been settled.

By _incomplete solution_, I mean just specifying `Library = "msl"` on the external function, and on the library level describe a `SourceDirectory` for the `msl` library.  Details of how that library should be built can be left to #1668.  Sure, this isn't how things will work in MLS 3.5, but I suppose I'm not the only one who can see this becoming the natural way of doing it once we have a way to describe building of libraries.  If we can see that specifying `SourceDirectory` on every function is going to become an obsolete technique, it seems wise to compare the added value of specifying it to the confusion resulting from adopting a method already expected to become obsolete.

**Comment by henrikt-ma** [comment]
> The include annotation is in a pretty weird area. It's not necessary to add an include for the prototypes since the specification says exactly what they will look like. And tools that do not compile code (only performing dynamic loading of libraries), cannot use it.

For tools that can use it, it gives valuable protection against programs crashing because the Modelica external function wrapper call isn't matching the implementation (assuming that the library was compiled with the same header included).  For this reason, I consider it bad style to rely on automatic generation of prototypes (which is what you get when no `Include` is given).

**Comment by sjoelund** [comment]
> By _incomplete solution_, I mean just specifying `Library = "msl"` on the external function, and on the library level describe a `SourceDirectory` for the `msl` library.

I'll assume this was just to be considered an example since MSL has multiple libraries (one of which is ModelicaExternalC).

**Comment by sjoelund** [comment]
> For tools that can use it, it gives valuable protection against programs crashing because the Modelica external function wrapper call isn't matching the implementation (assuming that the library was compiled with the same header included). For this reason, I consider it bad style to rely on automatic generation of prototypes (which is what you get when no `Include` is given).

I consider it bad style that you can include a header and get different datatypes as inputs to the function than what the specification says. And if you have both the default and the included function, it will fail if there is a static or inline specification for it.

**Comment by henrikt-ma** [comment]
> > By _incomplete solution_, I mean just specifying `Library = "msl"` on the external function, and on the library level describe a `SourceDirectory` for the `msl` library.
> 
> I'll assume this was just to be considered an example since MSL has multiple libraries (one of which is ModelicaExternalC).

Yes, it was just an example I made up; `Library = "ModelicaExternalC"` would have been a better example.

**Comment by sjoelund** [comment]
> Speaking of which, how come this isn't present here (I'm not really familiar with the pre-built binaries, since we have a product that also needs to run on macOS)?
> https://github.com/modelica/ModelicaStandardLibrary/tree/master/Modelica/Resources/Library/linux64

I never liked to have them in git anyway. Would be much nicer with CI/CD deploying binaries or zipping up the library.

**Comment by beutlich** [comment]
> I never liked to have them in git anyway. Would be much nicer with CI/CD deploying binaries or zipping up the library.

I know. For that reason I try to update the binaries as rarely as possible.

**Comment by beutlich** [comment]
I'd really like to have this issue resolved/closed before the voting. When I opened it two weeks ago my intention was to have an example of the SourceDirectory annotation, too. If that is not going to happen, someone can close it.

**Comment by HansOlsson** [comment]
> I'd really like to have this issue resolved/closed before the voting. When I opened it two weeks ago my intention was to have an example of the SourceDirectory annotation, too. If that is not going to happen, someone can close it.

To me adding an example wasn't that high priority so I didn't prepare a PR in time for the vote. It might also be that there were to me this new discussion that I missed.
However, I believe it can be added in the next release.

**Comment by beutlich** [comment]
This issue is still open while targeting the closed milestone 3.5. I propose to first get it resolved before the MA approval of MLS 3.5.

**Comment by HansOlsson** [comment]
> This issue is still open while targeting the closed milestone 3.5. I propose to first get it resolved before the MA approval of MLS 3.5.

It was re-opened with the intent of adding an example.
I don't see adding an example as that blocking, even if it would have been good.

**Comment by HansOlsson** [comment]
> Just a question: Could we also update the example in 12.9.4 to make use of the SourceDirectory annotation. I was thinking of moving ExternalFunc3.c from Include to Sources, which would require to change the IncludeDirectory of ExternalFunc3, too.
> 
> Reopening to just not forget it.

I believe it is cleanest to add a new ticket for the missing example: #2847, and close this one.
